### ISSUE TITLE
[UI] Dark mode

### ISSUE THREAD

**Issue by jstanden** [jstanden/cerb#1281]
We'll gather dark mode requests here.

At the moment, the request is white text on black background, and checking the browser/OS preference.

**Comment by jstanden** [comment]
There's a custom stylesheet in Setup->Configure->Branding if someone wants to tinker with this.

**Comment by iammattmartin** [comment]
We did some of these in a quick and nasty change.

```
HTML, BODY {
    background-color: #171717;
}

BODY, FORM, TD {
    color: #fff;
}

A {
    color: #d9d9d9;
}

.ui-tabs .ui-tabs-nav {
    background: #1f1f1f none;
}

.ui-widget-content {
    background: #1f1f1f none;
}

.ui-tabs .ui-tabs-panel {
    background: #1f1f1f none;
}

.ui-widget-content a {
    color: #a4a4a4;
}


.ui-tabs .ui-tabs-nav li {
    background: -moz-linear-gradient(270deg, #4a4a4a, #000);
    border-color: #5b5b5b;
}

.ui-tabs .ui-tabs-nav li.ui-state-default a {
    color: #ececec;
}

.ui-tabs .ui-tabs-nav li.ui-tabs-active {
    background: -moz-linear-gradient(270deg, #60afdf, #090909);
    border-color: #5b5b5b;
}

.ui-tabs .ui-tabs-nav li.ui-tabs-active a {
    color: #fff;
}

.glyphicons {
    color: white;
}

TABLE.worklistBody TR.even {
    background-color: #2b2b2b;
}

TABLE.worklistBody TR.odd {
    background-color: #222;
}

DIV.badge {
    background: -moz-linear-gradient(270deg, #807f7f, #3a3a3a);
}

DIV.badge-lightgray {
    background: -moz-linear-gradient(270deg, #313131, #464646);
    color: #fff;
}

.ui-widget-content {
    color: #f2f2f2;
}

TABLE.worklistBody TR.selected {
    background-color: #10acf957 !important;
}

TABLE.worklistBody TR.hover {
    background-color: #00abff82 !important;
}

TABLE.worklistBody TR.selected.hover {
    background-color: #19acf482 !important;
}

TABLE.worklistBody TD A.subject, TABLE.worklistBody TD B.subject {
    color: #bdd8e8;
}

.ui-tabs .ui-tabs-nav li.ui-state-hover {
    background: -moz-linear-gradient(270deg, #a2cbe3, #606060);
}

TABLE.worklistBody th {
    background: -moz-linear-gradient(270deg, #515151, #262626);
}

TABLE.worklistBody th A {
    color: #dfdfdf;
}

UL.navmenu > LI {
    background-color: #3e3e3e;
}

UL.navmenu > LI A {
    color: #fff;
}

UL.cerb-popupmenu {
    background-color: #171717;
    box-shadow: 0px 0px 15px 0px #464646;
    border: 1px solid #717070;
}

.ui-dialog {
	box-shadow: 0 0 10px 3px #464646;
}

.cerb-subpage H1 {
    color: white;
}

.ui-dialog h1 {
    color: white;
}

DIV.block {
    background: -moz-linear-gradient(270deg, #282828 25px, #171717 75px);
    background-color: rgba(0, 0, 0, 0);
}

.ui-widget textarea {
   color: white;
   background-color: #171717;
}

FIELDSET.properties LEGEND {
   color: white;
}

FORM INPUT[type="text"], FORM INPUT[type="password"]  {
    border: 1px solid #313131;
    background-color: #2d2d2d;
    color: white;
}

.ui-widget input, .ui-widget select  {
    background-color: #2d2d2d;
    color: white;
}

.ace-cerb {
    background-color: #404040;
    color: white;
}

DIV.help-box {
   background: -moz-linear-gradient(270deg, #666141, #8a8a8a);
}

DIV.block {
    border: 1px solid #5b5b5b;
}

UL.navmenu > LI:hover {
    background-color: #5f82a8;
}

FIELDSET {
    border: 1px double #4f4f4f;
}

BUTTON {
    color: #d9d9d9;
    background: -moz-linear-gradient(270deg, #595959, #5c5c5c);
}

BUTTON:hover {
    color: #d5d5d5;
    background: -moz-linear-gradient(270deg, #8a8a8a, #757575);
	
}

.ace-cerb .ace_gutter {
    background: #3a3a3a;
}

BUTTON SPAN.glyphicons {
    color: #c2c2c2;
}

UL.bubbles.chooser-container LI A {
    color: #fbfbfb;
}

UL.bubbles.chooser-container LI {
	background: -moz-linear-gradient(270deg, #313131, #646464);
	color: #aaa;
}

UL.bubbles LI {
   border: 1px solid #464646;
}

TABLE.cerb-widget-data-table tbody:nth-child(2n) td {
    background-color: #393939;
}

UL.bubbles LI.bubble-gray A {
    color: white;
}

UL.bubbles LI.bubble-gray {
	background: -moz-linear-gradient(270deg, #313131, #646464);
	border: 1px solid #464646;
	color: white;
}

UL.bubbles LI.bubble-blue {
  	border: 1px solid #555;
   color: #5ea1ea;
   background: -moz-linear-gradient(270deg, #666, #404040);
}

.ui-dialog .ui-dialog-titlebar {
    background-color: #4c5d88;
}

DIV.cerb-menu > UL > LI {
    background-color: #4c5d88;
}

UL.bubbles LI A {
    color: #e5e5e5;
}

UL.bubbles LI {
   background: -moz-linear-gradient(270deg, #666, #404040);
}

.cerb-code-editor-toolbar {
    background: #262626;
    border-bottom: 1px solid #3e3d3d;
    border-top: 1px solid #3e3d3d;
}

.cerb-code-editor-toolbar .cerb-code-editor-toolbar-button {
    background: #4f4d4d;
}
```

Edit: few buttons added

**Comment by jstanden** [comment]
That's a great start! It looks like it's mostly just the jQueryUI styles left.

We have the Sass templates here:
https://github.com/jstanden/cerb/tree/9.5.10/install/extras/developers/css/cerb.css

I could refactor that a bit so the colors (theme) are separate from the layout. Then we could move the above into a dark theme. We'd probably just generate a `cerb.css` and `cerb-dark.css` and have worker prefs switch between them.

**Comment by iammattmartin** [comment]
I've added a few more bits as ours needed some changes.

There will be the odd bit left out so will update as I can.

Scheduler page seems to have hard-set elements which give a white background.
Translations page seems to have hard-set elements which give dark text.
Ace editor seems to have a white line selected element which can't yet find to make darker.
The blue bar across the top appears to be hard-set.

**Comment by jstanden** [comment]
Yeah, there are quite a few `style` attributes that we need to clean up. Now that 9.6 is released I should have some time to do that cleanup. The 10.0 branch is a good place since it's a rare opportunity to deprecate/retire parts of the old CSS classname hierarchy.

**Comment by jstanden** [comment]
This still has a reasonable amount of interest. We need to finish up 10.0 so we have time to spend on the Sass refactor for separating layout and themes.

**Comment by bwoodruff** [comment]
Adding a +1 here now that 10.0 is out

**Comment by DaftBrit** [comment]
+1 on this from me :)

**Comment by claphamj** [comment]
Dark mode would make Cerb the best ;)

**Comment by triggeroni** [comment]
Hi Folks, my eyes are burning when using Cerb, and it is such a vital tool for our operation, I fully support the coding for Dark Mode ASAP!

**Comment by jstanden** [comment]
Alright! I've (finally) resumed work on this in 10.2. We're migrating to CSS variables (in Sass) to make theming easier.

I'm cleaning up the templates so there aren't any explicit colors in styles. Instead, they're variables like `var(--cerb-color-button-text)`. We can then override those variables with a `.dark` class on something like `<BODY>` -- which could be dynamically toggled in the UI.

It'll be pretty quick to cover 90% of the UI styles, but there are literally a few thousand inline `style` attributes in our templates that need to be cleaned up. Most of those wouldn't affect themes (e.g. `float: right` and `display: none`). I imagine we can automatically clean up the majority of them.

I could use some input from the more aesthetically inclined when we're ready to define the actual colors on dark mode.

We're already rebasing 10.2 to do a preview release for some other functionality. I'll include this in a preview release so you all can help tune it.

**Comment by DaftBrit** [comment]
Thanks for looking into this @jstanden. In terms of pallets the discord messaging app has to be one of favorite contrasts: [Discord Branding](https://www.onlinepalette.com/wp-content/uploads/2019/12/discord-source.png)

**Comment by jstanden** [comment]
We finished going through all of our Saas/SCSS templates and separating out colors into abstract CSS variables.

Colors can now be tweaked from a single place on a few lines. We can deploy a couple themes (light/dark), and those can be overridden from a custom stylesheet (Setup->Configure->Branding).

I did a quick/dirty dark mode by inverting most of the colors and it works well. We just need to finish overriding the styles for a few UI libraries so they properly inherit the color preferences.

For instance, the defaults look like:

``` .css
:root {
  --cerb-font-size: 12px;
  --cerb-font-stack: "Helvetica Neue", Helvetica, Arial, sans-serif;
  --cerb-font-stack-fixed: Lucida Console, Lucida Sans Typewriter, monaco, Bitstream Vera Sans Mono, monospace;
  
  --cerb-color-text: rgb(0,0,0);
  --cerb-color-text-inverse: rgb(255,255,255);
  
  --cerb-color-background: rgb(255,255,255);
  --cerb-color-background-contrast-250: rgb(250,250,250);
  --cerb-color-background-contrast-240: rgb(240,240,240);
  --cerb-color-background-contrast-230: rgb(230,230,230);
  --cerb-color-background-contrast-220: rgb(220,220,220);
  --cerb-color-background-contrast-210: rgb(210,210,210);
  --cerb-color-background-contrast-200: rgb(200,200,200);
  --cerb-color-background-contrast-190: rgb(190,190,190);
  --cerb-color-background-contrast-180: rgb(180,180,180);
  --cerb-color-background-contrast-175: rgb(175,175,175);
  --cerb-color-background-contrast-165: rgb(165,165,165);
  --cerb-color-background-contrast-150: rgb(150,150,150);
  --cerb-color-background-contrast-125: rgb(125,125,125);
  --cerb-color-background-contrast-100: rgb(100,100,100);
  --cerb-color-background-contrast-75: rgb(75,75,75);
  --cerb-color-background-contrast-50: rgb(50,50,50);
  --cerb-color-background-contrast-25: rgb(25,25,25);
  --cerb-color-background-inverse: rgb(0,0,0);
//...
```

Those abstract names represent the degree of contrast from the background (e.g. `250/255` for light, or `(255-250)/255` for dark). We converted all of the `rgb()` references to `var(--cerb-color-*)`. We're still in the process of consolidating that into fewer shades.

We did the same thing with other controls like menus, buttons, headings, etc.

The dark mode overrides that like:

``` .css
.dark {
  --cerb-color-text: rgb(255,255,255);
  --cerb-color-text-inverse: rgb(0,0,0);

  --cerb-color-background: rgb(0,0,0);
  --cerb-color-background-contrast-250: rgb(5,5,5);
  --cerb-color-background-contrast-240: rgb(15,15,15);
  --cerb-color-background-contrast-230: rgb(25,25,25);
  --cerb-color-background-contrast-220: rgb(35,35,35);
  --cerb-color-background-contrast-210: rgb(45,45,45);
  --cerb-color-background-contrast-200: rgb(55,55,55);
  --cerb-color-background-contrast-190: rgb(65,65,65);
  --cerb-color-background-contrast-180: rgb(75,75,75);
  --cerb-color-background-contrast-175: rgb(80,80,80);
  --cerb-color-background-contrast-165: rgb(90,90,90);
  --cerb-color-background-contrast-150: rgb(95,95,95);
  --cerb-color-background-contrast-125: rgb(130,130,130);
  --cerb-color-background-contrast-100: rgb(155,155,155);
  --cerb-color-background-contrast-75: rgb(180,180,180);
  --cerb-color-background-contrast-50: rgb(205,205,205);
  --cerb-color-background-contrast-25: rgb(233,233,233);
  --cerb-color-background-inverse: rgb(255,255,255);
//...
```

The current light theme:

<img width="1360" alt="image" src="https://user-images.githubusercontent.com/63288/135951462-e79ef560-0b9c-49a6-8ec2-37ede964aa35.png">

If we add the `dark` class to the HTML document (which can happen dynamically in the UI with one click):

``` .html
<html class="dark">
```

We get (the currently very rough dark theme):

<img width="1360" alt="image" src="https://user-images.githubusercontent.com/63288/135951522-4eca59d8-f29e-4e14-a4c3-6e3413f050ea.png">

That obviously needs some tweaks aesthetically (the tabs, notifications, etc) -- but the dark mode PoC was only about 15 mins of work once the styles were cleaned up.

This should make it pretty easy for a proper designer to work with.

**Comment by jstanden** [comment]
More progress on overriding third-party UI components:

<img width="1470" alt="image" src="https://user-images.githubusercontent.com/63288/136112786-fb8a5846-7284-4879-a7f4-e860d4e3501d.png">

**Comment by jstanden** [comment]
This is included in 10.2, but the stylesheet could still use some tuning from a proper designer.
